File: web/packages/teleport/src/services/databases/databases.test.ts
/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import api from 'teleport/services/api';

import DatabaseService from './databases';
import { Database, IamPolicyStatus } from './types';

test('correct formatting of database fetch response', async () => {
  jest.spyOn(api, 'get').mockResolvedValue(mockResponse);

  const database = new DatabaseService();
  const response = await database.fetchDatabases('im-a-cluster', {
    search: 'does-not-matter',
  });

  expect(response).toEqual({
    agents: [
      {
        kind: 'db',
        name: 'aurora',
        description: 'PostgreSQL 11.6: AWS Aurora',
        type: 'Amazon RDS PostgreSQL',
        protocol: 'postgres',
        names: [],
        users: [],
        roles: [],
        hostname: '',
        labels: [
          { name: 'cluster', value: 'root' },
          { name: 'env', value: 'aws' },
        ],
        aws: {
          rds: {
            resourceId: 'resource-id',
            region: 'us-west-1',
            vpcId: 'vpc-123',
            subnets: ['sn1', 'sn2'],
            securityGroups: [],
          },
          iamPolicyStatus: IamPolicyStatus.Success,
        },
        supportsInteractive: false,
      },
      {
        kind: 'db',
        name: 'self-hosted',
        type: 'Self-hosted PostgreSQL',
        protocol: 'postgres',
        names: [],
        users: [],
        roles: [],
        hostname: '',
        labels: [],
        aws: undefined,
        supportsInteractive: true,
      },
    ],
    startKey: mockResponse.startKey,
    totalCount: mockResponse.totalCount,
  });
});

test('null response from database fetch', async () => {
  jest.spyOn(api, 'get').mockResolvedValue(null);

  const database = new DatabaseService();
  const response = await database.fetchDatabases('im-a-cluster', {
    search: 'does-not-matter',
  });

  expect(response).toEqual({
    agents: [],
    startKey: undefined,
    totalCount: undefined,
  });
});

describe('correct formatting of all type and protocol combos', () => {
  test.each`
    type                     | protocol                 | combined
    ${'self-hosted'}         | ${'mysql'}               | ${'Self-hosted MySQL/MariaDB'}
    ${'rds'}                 | ${'mysql'}               | ${'Amazon RDS MySQL/MariaDB'}
    ${'self-hosted'}         | ${'postgres'}            | ${'Self-hosted PostgreSQL'}
    ${'rds'}                 | ${'postgres'}            | ${'Amazon RDS PostgreSQL'}
    ${'rdsproxy'}            | ${'sqlserver'}           | ${'Amazon RDS Proxy SQL Server'}
    ${'gcp'}                 | ${'postgres'}            | ${'Cloud SQL PostgreSQL'}
    ${'redshift'}            | ${'postgres'}            | ${'Amazon Redshift'}
    ${'redshift-serverless'} | ${'postgres'}            | ${'Amazon Redshift Serverless'}
    ${'dynamodb'}            | ${'dynamodb'}            | ${'Amazon DynamoDB'}
    ${'elasticache'}         | ${'redis'}               | ${'Amazon ElastiCache'}
    ${'memorydb'}            | ${'redis'}               | ${'Amazon MemoryDB'}
    ${'opensearch'}          | ${'opensearch'}          | ${'Amazon OpenSearch'}
    ${'keyspace'}            | ${'cassandra'}           | ${'Amazon Keyspaces'}
    ${'self-hosted'}         | ${'sqlserver'}           | ${'Self-hosted SQL Server'}
    ${'self-hosted'}         | ${'redis'}               | ${'Self-hosted Redis'}
    ${'self-hosted'}         | ${'mongodb'}             | ${'Self-hosted MongoDB'}
    ${'self-hosted'}         | ${'cassandra'}           | ${'Self-hosted Cassandra'}
    ${'self-hosted'}         | ${'cockroachdb'}         | ${'Self-hosted CockroachDB'}
    ${'self-hosted'}         | ${'oracle'}              | ${'Self-hosted Oracle'}
    ${'self-hosted'}         | ${'snowflake'}           | ${'Snowflake'}
    ${'self-hosted'}         | ${'elasticsearch'}       | ${'Self-hosted Elasticsearch'}
    ${'some other type'}     | ${'some other protocol'} | ${'some other type some other protocol'}
  `(
    'should combine type: $type and protocol: $protocol correctly',
    async ({ type, protocol, combined }) => {
      jest.spyOn(api, 'get').mockResolvedValue({ items: [{ type, protocol }] });

      const database = new DatabaseService();
      const response = await database.fetchDatabases('im-a-cluster', {
        search: 'does-not-matter',
      });

      const dbs = response.agents as Database[];
      expect(dbs[0].type).toBe(combined);
    }
  );
});

test('null labels field in database fetch response', async () => {
  jest.spyOn(api, 'get').mockResolvedValue({ items: [{ labels: null }] });

  const database = new DatabaseService();
  const response = await database.fetchDatabases('im-a-cluster', {
    search: 'does-not-matter',
  });

  expect(response.agents[0].labels).toEqual([]);
});

test('database services fetch response', async () => {
  const database = new DatabaseService();

  jest.spyOn(api, 'get').mockResolvedValue(mockServiceResponse);
  const response = await database.fetchDatabaseServices('im-a-cluster');
  expect(response.services).toEqual([
    {
      name: 'db-service-1',
      matcherLabels: {
        env: ['prod', 'env'],
        os: ['mac', 'ios', 'linux', 'windows'],
        tag: ['tag'],
        fruit: ['apple'],
      },
    },
    {
      name: 'db-service-2',
      matcherLabels: {},
    },
  ]);
});

test('null array fields in database services fetch response', async () => {
  const database = new DatabaseService();

  jest.spyOn(api, 'get').mockResolvedValue({});
  let response = await database.fetchDatabaseServices('im-a-cluster');
  expect(response.services).toEqual([]);

  jest.spyOn(api, 'get').mockResolvedValue({ items: [{ name: '' }] });
  response = await database.fetchDatabaseServices('im-a-cluster');
  expect(response.services).toEqual([{ name: '', matcherLabels: {} }]);
});

const mockResponse = {
  items: [
    // aws rds
    {
      name: 'aurora',
      desc: 'PostgreSQL 11.6: AWS Aurora',
      protocol: 'postgres',
      type: 'rds',
      uri: 'postgres-aurora-instance-1.c1xpjrob56xs.us-west-1.rds.amazonaws.com:5432',
      hostname: '',
      labels: [
        { name: 'cluster', value: 'root' },
        { name: 'env', value: 'aws' },
      ],
      aws: {
        rds: {
          resource_id: 'resource-id',
          region: 'us-west-1',
          vpc_id: 'vpc-123',
          subnets: ['sn1', 'sn2'],
          security_groups: [],
        },
        iam_policy_status: 'IAM_POLICY_STATUS_SUCCESS',
      },
    },
    // non-aws self-hosted
    {
      name: 'self-hosted',
      type: 'self-hosted',
      protocol: 'postgres',
      uri: 'localhost:5432',
      labels: [],
      hostname: '',
      supports_interactive: true,
    },
  ],
  startKey: 'mockKey',
  totalCount: 100,
};

const mockServiceResponse = {
  items: [
    {
      name: 'db-service-1',
      resource_matchers: [
        {
          labels: { env: ['prod', 'env'], os: ['mac', 'ios'], fruit: 'apple' },
        },
        { labels: { os: ['linux', 'windows'], tag: ['tag'] } },
      ],
    },
    {
      name: 'db-service-2',
      resource_matchers: [],
    },
  ],
  startKey: 'mockKey',
  totalCount: 100,
};

File: lib/services/local/databases_test.go
/*
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package local

import (
	"context"
	"testing"

	"github.com/google/go-cmp/cmp"
	"github.com/google/go-cmp/cmp/cmpopts"
	"github.com/gravitational/trace"
	"github.com/jonboulle/clockwork"
	"github.com/stretchr/testify/require"

	"github.com/gravitational/teleport/api/types"
	"github.com/gravitational/teleport/lib/backend/memory"
	"github.com/gravitational/teleport/lib/defaults"
)

// TestDatabasesCRUD tests backend operations with database resources.
func TestDatabasesCRUD(t *testing.T) {
	ctx := context.Background()

	backend, err := memory.New(memory.Config{
		Context: ctx,
		Clock:   clockwork.NewFakeClock(),
	})
	require.NoError(t, err)

	service := NewDatabasesService(backend)

	// Create a couple databases.
	db1, err := types.NewDatabaseV3(types.Metadata{
		Name: "db1",
	}, types.DatabaseSpecV3{
		Protocol: defaults.ProtocolPostgres,
		URI:      "localhost:5432",
	})
	require.NoError(t, err)
	db2, err := types.NewDatabaseV3(types.Metadata{
		Name: "db2",
	}, types.DatabaseSpecV3{
		Protocol: defaults.ProtocolMySQL,
		URI:      "localhost:3306",
	})
	require.NoError(t, err)

	// Initially we expect no databases.
	out, err := service.GetDatabases(ctx)
	require.NoError(t, err)
	require.Empty(t, out)

	// Create both databases.
	err = service.CreateDatabase(ctx, db1)
	require.NoError(t, err)
	err = service.CreateDatabase(ctx, db2)
	require.NoError(t, err)

	// Try to create an invalid database.
	dbBadURI, err := types.NewDatabaseV3(types.Metadata{
		Name: "db-missing-port",
	}, types.DatabaseSpecV3{
		Protocol: defaults.ProtocolMySQL,
		URI:      "localhost",
	})
	require.NoError(t, err)
	require.NoError(t, service.CreateDatabase(ctx, dbBadURI))

	// Fetch all databases.
	out, err = service.GetDatabases(ctx)
	require.NoError(t, err)
	require.Empty(t, cmp.Diff([]types.Database{dbBadURI, db1, db2}, out,
		cmpopts.IgnoreFields(types.Metadata{}, "ID", "Revision"),
	))

	// Fetch a specific database.
	db, err := service.GetDatabase(ctx, db2.GetName())
	require.NoError(t, err)
	require.Empty(t, cmp.Diff(db2, db,
		cmpopts.IgnoreFields(types.Metadata{}, "ID", "Revision"),
	))

	// Try to fetch a database that doesn't exist.
	_, err = service.GetDatabase(ctx, "doesnotexist")
	require.IsType(t, trace.NotFound(""), err)

	// Try to create the same database.
	err = service.CreateDatabase(ctx, db1)
	require.IsType(t, trace.AlreadyExists(""), err)

	// Update a database.
	db1.Metadata.Description = "description"
	err = service.UpdateDatabase(ctx, db1)
	require.NoError(t, err)
	db, err = service.GetDatabase(ctx, db1.GetName())
	require.NoError(t, err)
	require.Empty(t, cmp.Diff(db1, db,
		cmpopts.IgnoreFields(types.Metadata{}, "ID", "Revision"),
	))

	// Delete a database.
	err = service.DeleteDatabase(ctx, db1.GetName())
	require.NoError(t, err)
	out, err = service.GetDatabases(ctx)
	require.NoError(t, err)
	require.Empty(t, cmp.Diff([]types.Database{dbBadURI, db2}, out,
		cmpopts.IgnoreFields(types.Metadata{}, "ID", "Revision"),
	))

	// Try to delete a database that doesn't exist.
	err = service.DeleteDatabase(ctx, "doesnotexist")
	require.IsType(t, trace.NotFound(""), err)

	// Delete all databases.
	err = service.DeleteAllDatabases(ctx)
	require.NoError(t, err)
	out, err = service.GetDatabases(ctx)
	require.NoError(t, err)
	require.Empty(t, out)
}
